vm/devices/net/net_mana/src/lib.rs

```diff
@@ -1118,18 +1119,94 @@ impl<T: DeviceBacking> ManaQueue<T> {
+                while segments_remaining >= 31 && segments_peekable.peek().is_some() {
+                    let seg = segments_peekable.next().unwrap();
+                    let next_segment_len = if let Some(&next) = segments_peekable.peek() {
+                        next.len
+                    } else {
+                        0
+                    };
+                    // NOTE: allocate will panic if size is larger than PAGE_SIZE32 (4096)
+                    if segments_peekable.peek().is_none()
+                        || seg.len + next_segment_len >= PAGE_SIZE32
+                    {
+                        sgl[sgl_index] = Sge {
+                            address: self.guest_memory.iova(seg.gpa).unwrap(),
+                            mem_key: self.mem_key,
+                            size: seg.len,
+                        };
+                        sgl_index += 1;
+                        segments_remaining -= 1;
+                        segment_index += 1;
+                    } else {
+                        // Combine segments and bounce buffer
+                        let buffer_len = seg.len + next_segment_len;
+                        let mut buf: ContiguousBuffer<'_> =
+                            match self.tx_bounce_buffer.allocate(buffer_len as u32) {
+                                Ok(buf) => buf,
+                                Err(err) => {
+                                    tracelimit::error_ratelimited!(
+                                        err = &err as &dyn std::error::Error,
+                                        meta.len,
+                                        "failed to bounce buffer"
+                                    );
+                                    // Drop the packet
+                                    return Ok(None);
+                                }
+                            };
+                        let mut next = buf.as_slice();
+                        let len = seg.len as usize;
+                        self.guest_memory
+                            .read_to_atomic(seg.gpa as u64, &next[..len])?;
+                        next = &next[len..];
+                        let seg = segments_peekable.next().unwrap();
+                        let len = seg.len as usize;
+                        self.guest_memory
+                            .read_to_atomic(seg.gpa as u64, &next[..len])?;
+
+                        let buf = buf.commit();
+                        sgl[sgl_index] = Sge {
+                            address: buf.gpa,
+                            mem_key: self.mem_key,
+                            size: buffer_len as u32,
+                        };
+                        sgl_index += 1;
+                        segments_remaining -= 2;
+                        segment_index += 2;
+                        bounced_len_with_padding_total += buffer_len;
+                    }
+                }
+                let sgl = &mut sgl[..segments_remaining];
```

Comment:
I am not convinced that this is still less than 31. If you started with 1000 segments, for example (I know that's probably not possible, but bear with me), then it seems that the loop above could at best get you down to 500 segments.